renderer/Profiling.cpp

```diff
@@ -0,0 +1,181 @@
+/*****************************************************************************
+The Dark Mod GPL Source Code
+
+This file is part of the The Dark Mod Source Code, originally based
+on the Doom 3 GPL Source Code as published in 2011.
+
+The Dark Mod Source Code is free software: you can redistribute it
+and/or modify it under the terms of the GNU General Public License as
+published by the Free Software Foundation, either version 3 of the License,
+or (at your option) any later version. For details, see LICENSE.TXT.
+
+Project: The Dark Mod (http://www.thedarkmod.com/)
+
+******************************************************************************/
+#include "precompiled.h"
+#pragma hdrstop
+
+#include "Profiling.h"
+
+idCVar r_useDebugGroups( "r_useDebugGroups", "0", CVAR_RENDERER | CVAR_BOOL, "Emit GL debug groups during rendering. Useful for frame debugging and analysis with e.g. nSight, which will group render calls accordingly." );
+idCVar r_glProfiling( "r_glProfiling", "0", CVAR_RENDERER | CVAR_BOOL, "Collect profiling information about GPU and CPU time spent in the rendering backend." );
+
+/**
+ * Tracks CPU and GPU time spent in profiled sections.
+ * Sections are structured in hierarchical fashion, you can push and pop sections onto a stack.
+ * Each section has a name; sections with the same name at the same place in the hierarchy will
+ * be accumulated to a single total time in the end.
+ */
+class GlProfiler {
+public:
+	GlProfiler() : profilingActive(false), frameMarker(0) {}
+
+	void BeginFrame() {
+		sectionStack.clear();
+		frame[ frameMarker ] = section();
+		frame[ frameMarker ].name = "Frame";
```

Comment:
Maybe add frame index here (I guess we have some numbering of frames)...

Reply:
We do, but I'm not sure how readily it is accessible from this part of the code. If it's not too invasive, I'll add it.

Reply:
Btw, I can't seem to answer to your glext question directly, so I'll put it here: I had to update the glext.h with a newer version directly from the Khronos group. The reason is that our current version is at an age where it does not contain any GL4 stuff, so the glPushDebugGroup and glPopDebugGroup functions are missing. The update shouldn't break anything, it's just updated extension definitions, basically.

Reply:
Well, unless previous version had local modification...
Oh God! How could I even think of it =)

Reply:
Not very likely, and if so, all the more reason to clean it up :D
But again, the glext.h only contains type information about all available extension methods, it does not actually provide any implementations. So the very fact that TDM still compiles with the new version almost certainly guarantees that the update didn't break anything.